core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala

```diff
@@ -414,6 +433,142 @@ private[spark] object ClosureCleaner extends Logging {
+
+  /**
+   * Check if the given reference is a indylambda style Scala closure.
+   * If so, return a non-empty serialization proxy (SerializedLambda) of the closure;
+   * otherwise return None.
+   *
+   * @param maybeClosure the closure to check.
+   */
+  def getSerializationProxy(maybeClosure: AnyRef): Option[SerializedLambda] = {
+    val maybeClosureClass = maybeClosure.getClass
+
+    // shortcut the fast check:
+    // indylambda closure classes are generated by Java's LambdaMetafactory, and they're always
+    // synthetic.
+    if (!maybeClosureClass.isSynthetic) return None
+
+    val implementedInterfaces = ClassUtils.getAllInterfaces(maybeClosureClass).asScala
+    val isClosureCandidate = implementedInterfaces.exists(_.getName == "scala.Serializable") &&
+      implementedInterfaces.exists(_.getName.startsWith("scala.Function"))
+
+    if (isClosureCandidate) {
+      try {
+        val lambdaProxy = inspect(maybeClosure)
+        if (isIndylambdaScalaClosure(lambdaProxy)) Option(lambdaProxy)
+        else None
+      } catch {
+        case e: Exception =>
+          // no need to check if debug is enabled here the Spark logging api covers this.
+          logDebug("The given reference is not an indylambda Scala closure.", e)
+          None
+      }
+    } else {
+      None
+    }
+  }
+
+  def isIndylambdaScalaClosure(lambdaProxy: SerializedLambda): Boolean = {
+    lambdaProxy.getImplMethodKind == MethodHandleInfo.REF_invokeStatic &&
+      lambdaProxy.getImplMethodName.contains("$anonfun$")
+      // && implements a scala.runtime.java8 functional interface
+  }
+
+  def inspect(closure: AnyRef): SerializedLambda = {
+    val writeReplace = closure.getClass.getDeclaredMethod("writeReplace")
+    writeReplace.setAccessible(true)
+    writeReplace.invoke(closure).asInstanceOf[SerializedLambda]
+  }
+
+  /**
+   * Scans an indylambda Scala closure, along with its lexically nested closures, and populate
+   * the accessed fields info on which fields on the outer object are accessed.
+   */
+  def findAccessedFields(
+      lambdaProxy: SerializedLambda,
+      lambdaClassLoader: ClassLoader,
+      accessedFields: Map[Class[_], Set[String]]): Unit = {
+    val implClassInternalName = lambdaProxy.getImplClass
+    // scalastyle:off classforname
+    val implClass = Class.forName(
+      implClassInternalName.replace('/', '.'), false, lambdaClassLoader)
+    // scalastyle:on classforname
+    val implClassNode = new ClassNode()
+    val implClassReader = ClosureCleaner.getClassReader(implClass)
+    implClassReader.accept(implClassNode, 0)
+
+    val methodsByName = Map.empty[MethodIdentifier[_], MethodNode]
+    for (m <- implClassNode.methods.asScala) {
+      methodsByName(MethodIdentifier(implClass, m.name, m.desc)) = m
+    }
+
+    val implMethodId = MethodIdentifier(
+      implClass, lambdaProxy.getImplMethodName, lambdaProxy.getImplMethodSignature)
+
+    val visited = Set[MethodIdentifier[_]](implMethodId)
```

Comment:
`visited` is not used.

Reply:
Oh shoot, I forgot to add the uses..... thanks for catching this!
It should be added right after the `val currentId = stack.pop` line. The initialization line should probably be empty instead of pre-populating with the `implMethodId`, for consistency.